IECoreImage : A new module for image processing

andrewkaufman opened the pull request:
This sprawling PR begins our efforts towards more modular sub-libraries, which is aimed at both improving discover-ability of code (by reducing files per include directory) and insulating external dependencies specific to sub-domains (e.g. the OpenImageIO dependency I'm introducing here).

The result is a new libIECoreImage (and associated IECoreImage python module) which contains all image processing code in Cortex (from ImagePrimitive itself, to its associated parameters and ops for reading/writing and manipulating images). libIECoreGL, libIECoreMaya, and libIECoreNuke all depend on libIECoreImage (as does libIECoreRI if you have SX enabled).

All format specific image readers/writers have been replaced with a single ImageReader and ImageWriter, both of which are backed by OpenImageIO. Furthermore, all color conversion and color processing has been removed from Cortex, as has support for deep image reading/writing. Many image related ops have been removed as well, on the grounds that other projects are better suited to those tasks (e.g. OCIO, OIIO, or Gaffer).

In general, IECoreImage should only be used for quite rudimentary image processing, such as loading textures in IECoreGL. Gaffer has much more performant (and well rounded) image processing capabilities.

It might be quite difficult to review at first, since so many files have been touched. I'll try to point out some helpful hints / drawbacks to watch for :

- 4c7a17e is the first commit that's not just removing (or adapting to recently removed) stuff
- 71cba77 duplicates some freetype code from `IECore::Font` to `IECoreImage::Font`, and `IECoreGL::Font` now has a pointer to both. I wasn't sure it'd be worth refactoring the freetype stuff to allow for something better.
- As discussed offline, `ImagePrimitive` is no longer a `Primitive`, and just a `BlindDataHolder`, though I'm still calling it `ImagePrimitive`. As a result, much of its API has changed in 35e2272.
- IECoreImage does depend on `CurvesPrimitive` right now (because of the `CurveThinner` op, moved in 48969f9). This isn't ideal, but I didn't want to lose the op, and I've left it for the future creator if IECoreScene to sort out.
- 7f3ea4f is the dodgiest commit in my books. It re-writes a bunch of test tifs (from testGL) using oiiotool. I don't know why those tiffs were crashing (the tiffs in testImage work just fine) and simply re-writing them fixes all the crashes, and the tests all pass unchanged (except one, addressed in 706ba02, but that wasn't passing before this PR either).
- `scons testImage` is not currently running on Travis, so the Travis tests only show that `testCore` and `testCorePython` are still passing. I have run `testImage` myself, in regular and debug mode, as well as `testGL`, `testRI`, `testArnold`, `testAppleseed`, `testMantra`, `testAlembic`, `testMaya`, `testNuke`, and `testHoudini`. That was all with gcc 4.8.3 on CentOs 7 at IE.
- The tests won't currently run at IE unless you have an env hack. That should be sorted out by IT shortly, but if anyone else did try to run them at IE right now, they'd crash.
- `scons test` does not currently run `testImage`. I'd like to fix that, but was unsure how.

johnhaddon commented:
I'm seeing some crashes running the tests at this end - I'll do some more digging. In the meantime, could you cherry-pick [this commit](https://github.com/johnhaddon/cortex/commit/493e2aa97ca2c311ba1aa5fd3b8eb901d2b9e929) over into your PR to fix the linking on OSX?

andrewkaufman commented:
Rebased on master and added John's osx linker fix.

andrewkaufman commented:
Quick update:

- `scons test` _does_  run testImage, provided you first `rm test/IECoreImage/results.txt`
- John fixed the failing Appleseed test on master, and it would likely be passing here as well, though I haven't tested yet because I need to recompile Appleseed at IE first (pointing to the correct OIIO).

johnhaddon commented:
What should I expect in terms of colourspace from the ImageReaders? If I recall correctly, before they used to guarantee the data was linear after loading, but that seems not to be the case now. That's meaning Gaffer's icons and such are all too dark...

andrewkaufman commented:
Currently I'm guaranteeing absolutely nothing about colorspace, I'm just returning whatever pixels OIIO hands me, optionally converted to float by OIIO. Though I see I left the doc string claiming a linear conversion... (https://github.com/ImageEngine/cortex/pull/605/commits/0bb799e6584f420e34e10379b17b8f8ee7e80c49#diff-ccdc5b7fcc977b0e98e94e612c3571dbR55). Is there a way to tell OIIO "you have to give me linear pixels"?

andrewkaufman commented:
I found this (quite old) [thread on the OIIO list](http://lists.openimageio.org/pipermail/oiio-dev-openimageio.org/2014-January/006481.html) saying its not gonna color convert from an `ImageInput`. So maybe I should mimic `oiiotool --autocc` in our `ImageReader`?

johnhaddon commented:
I'm not aware of a way of requesting linear pixels directly. There is `ImageBufAlgo::colorconvert()`, the `ColorProcessor` class and the other bits and bobs in `color.h` though, which you could use after loading. In Gaffer we originally attempted to guess the input colourspace using only the metadata that OIIO could load, but that didn't work out too well. We now use a config function which is configurable, but I think that configurability is overkill for the remaining ImageReader/ImageWriter uses. I'd suggest we borrow the logic from Gaffer's [default configuration](https://github.com/GafferHQ/gaffer/blob/908b4606bd55b51b49de98a4abf41e4118545487/startup/GafferImage/defaultColorSpace.py#L5) and use that in Cortex.

johnhaddon commented:
> So maybe I should mimic `oiiotool --autocc`

Oh, that one's new to me. Would be worth trying it, but from a very quick look I suspect it won't work in a lot of common cases (the one's that didn't work when we used the OIIO metadata in Gaffer).

andrewkaufman commented:
I pushed 2 new commits to try and address the colorspace conundrum. I'm certainly not confident that I've taken the right approach though, especially since I had to monkey with the JPG writing tests to get them to pass... I'm in the unfortunate situation of not being able to easily visualize any of this stuff until our OIIO install issues are sorted out.

> Would be worth trying it, but from a very quick look I suspect it won't work in a lot of common cases (the one's that didn't work when we used the OIIO metadata in Gaffer).

I have just gone with mimicking `oiiotool --autocc` for now, because its quite simple, and seems like it might work for the cases where `ImageReader` would normally be used (i.e. quite simple cases).

That default Gaffer color config seems to just hardcode some format based rules (e.g format x is in colorspace y), right? I don't particularly understand why those rules wouldn't be hardcoded into OIIO itself...  I really wanted OIIO to insulate me from this sort of thing... Maybe it'd be worth starting a conversation on their list about it?

andrewkaufman commented:
Ok, new attempt at naive color management here. The commits after c9e3b19 are all new. @johnhaddon let me know if the icons in Gaffer look alright now.

andrewkaufman commented on src/IECoreImage/ImageReader.cpp:
```diff
@@ -51,23 +55,336 @@ using namespace IECoreImage;
+		{
+			bool result = false;
+
+			if( ImageInput *input = ImageInput::open( filename ) )
```
If I recall, `create()` was not doing the necessary thing (e.g. failing on invalid images) so I switched to `open()`. I can try switching back to verify my memory.

andrewkaufman commented on src/IECoreImage/ImageWriter.cpp:
```diff
@@ -32,96 +32,534 @@
+	CompoundParameterPtr exrSettings = new CompoundParameter( "openexr", "OpenEXR specific settings" );
+	m_formatSettingsParameter->addParameter( exrSettings );
+	StringParameter::PresetsContainer exrCompressionPresets;
+	exrCompressionPresets.emplace_back( StringParameter::Preset( "none", "none" ) );
```
To be honest, I only switched to it because CLion told me to. That new stuff looks rather nice though, so I'll try that.

andrewkaufman commented:
> I get a crash every time. Can you repro that at your end? I'm building against Gaffer's latest dependencies packages on both platforms...

Nope, it never crashes at my end... I even upped the loop to 100000 and still no crash.

johnhaddon commented:
> Nope, it never crashes at my end... I even upped the loop to 100000 and still no crash.

Hmm. Interesting.  Do you know what version of libjpeg you're using? There's definitely something fishy going on, what with you needing to rewrite files to fix a crash at your end, and me having a crash on multiple platforms at this end. I guess next I should also try reverting those changed tiffs to see what happens here, and try reading that JPEG with gaffer and oiiotool...

andrewkaufman commented:
Its the CentOs 7 native one I believe: /usr/lib64/libjpeg.so.62.1.0

and from jconfig.h:

```
#define JPEG_LIB_VERSION 62
#define LIBJPEG_TURBO_VERSION 1.2.90
```

andrewkaufman commented on src/IECoreImage/ImageReader.cpp:
```diff
@@ -51,23 +55,336 @@ using namespace IECoreImage;
+		{
+			bool result = false;
+
+			if( ImageInput *input = ImageInput::open( filename ) )
```
Just tested again. If I use `create()` then `ImageReaderTest.testCanReadAndIsComplete()` passes, but `ImageReaderTest.testCannotReadDeepImage()` fails. It seems `create()` won't set the `ImageSpec.deep` flag for me...

andrewkaufman replied:
Addressed in 154f4304b88d0efbaccdf5f0c8da118dd541587f

andrewkaufman commented on src/IECoreImage/ImageReader.cpp:
```diff
@@ -51,23 +55,336 @@ using namespace IECoreImage;
+			{
+				// Reader::read() isn't const, but we know ours doesn't modify anything,
+				// so prefer a dodgy const_cast over replicating the functionality.
+				/// |todo: request a cheaper isComplete() mechanism in OpenImageIO
```
Addressed in 11f89726be98c57b186364e1b58ddf0806530582

andrewkaufman commented on src/IECoreImage/ImageReader.cpp:
```diff
@@ -51,23 +55,336 @@ using namespace IECoreImage;
+				auto cIt = members.find( thisName );
+
+				// create compound data
+				CompoundDataPtr newMetaData = nullptr;
```
Addressed in 77552770c5d9dd34106753ebf53dadf1593f12b0

andrewkaufman commented on src/IECoreImage/ImageReader.cpp:
```diff
@@ -51,23 +55,336 @@ using namespace IECoreImage;
+			m_inputFile = ImageInput::open( m_reader->fileName() );
+			if( !m_inputFile )
+			{
+				ImageInput::destroy( m_inputFile );
```
Addressed in b10e5080fc796749591b53462a5d9dbab476b491. Hopefully I used it right...

andrewkaufman commented on src/IECoreImage/ImageReader.cpp:
```diff
@@ -51,23 +55,336 @@ using namespace IECoreImage;
+	m_fileNameParameter->setTypedValue( fileName );
+}
+
+void ImageReader::constructCommon()
```
Addressed in b10e5080fc796749591b53462a5d9dbab476b491.

PS this is much nicer!

andrewkaufman commented on include/IECoreImage/ImageReader.h:
```diff
@@ -117,31 +107,27 @@ class IECOREIMAGE_API ImageReader : public IECore::Reader
 		IECore::BoolParameterPtr m_rawChannelsParameter;
+
+		IE_CORE_FORWARDDECLARE( Implementation );
+		ImplementationPtr m_implementation;
```
Addressed in b10e5080fc796749591b53462a5d9dbab476b491

andrewkaufman commented on include/IECoreImage/ImageWriter.h:
```diff
@@ -64,32 +66,43 @@ class IECOREIMAGE_API ImageWriter : public IECore::Writer
+
+		virtual ~ImageWriter();
+
+		/// Checks that object is an ImagePrimitive with consistent PrimitiveVariables
```
Addressed in ad25a99d4c76afbbb0732866b8073b754d5c3987

andrewkaufman replied:
Correction... addressed in 7ef1e3f93e541a184e43511910ec5ba5f71c2811

andrewkaufman commented on src/IECoreImage/ImageWriter.cpp:
```diff
@@ -32,96 +32,534 @@
+	CompoundParameterPtr exrSettings = new CompoundParameter( "openexr", "OpenEXR specific settings" );
+	m_formatSettingsParameter->addParameter( exrSettings );
+	StringParameter::PresetsContainer exrCompressionPresets;
+	exrCompressionPresets.emplace_back( StringParameter::Preset( "none", "none" ) );
```
Addressed in 7f00c5f5757f1e1f9c314141264fdbc82e77d917

andrewkaufman commented:
I think I've address all the comments so far, including the offline comment about needing access to a linear to sRGB conversion from python. That can now be done with `IECoreImage.ColorAlgo.transformImage( image, "linear", "sRGB" )`.

Still no idea why the JPG test is crashing for you. Its never crashed here. Also still would be nice to know why those tiff tests in `scons testGL` were crashing before I overwrote the files (or if they crash anywhere other than at Image Engine).

johnhaddon commented:
OK, progress. I've...

1. Run all the image tests on repeat with the truncated jpeg test disabled. All good, so it was just that one test causing problems, although sometimes the crashes actually occurred elsewhere.
2. Replaced my (Gaffer dependencies) libjpeg-8c with libjpeg-turbo-1.5.2. Ran all the image tests on repeat, including the truncated jpeg test. All good. So we can blame libjpeg for that one.
3. Ran the testGL tests, got no crashes.
4. Reverted the "rewrite testGL images" commit, got crashes.
5. Unreverted, and loaded every one of the old GL test tiffs in Gaffer via the ImageReader, without problem.

So the good news is the JPEG crash was a red herring, but the bad news is the TIFF crash manifests here too, and there's at least some evidence for it being a problem in the ImageReader...

andrewkaufman commented:
> So the good news is the JPEG crash was a red herring,

:+1: 

> the bad news is the TIFF crash manifests here too, and there's at least some evidence for it being a problem in the ImageReader...

:-1: 

Do you mind verifying with `LD_DEBUG=files` that you're getting the same libtiff and libOpenImageIO in Gaffer as you get with `scons testImage` ? I'll do the same here.

Does your crash also point to `LZWCleanup`?

Also, we just might have OIIO 1.8 installed here soon, and potentially OSL after that, so if you want to send a link to your Gaffer-uses-this-stuff branch, I can try that out here as well. If its not ready for PR yet, just send it privately.

johnhaddon commented:
> Do you mind verifying with LD_DEBUG=files that you're getting the same libtiff and libOpenImageIO in Gaffer as you get with scons testGL ?

Yep, same in each case. I also get a crash if I call `IECore.Reader.create( "crash.tif" ).read()` in Gaffer's ScriptEditor.

> Does your crash also point to LZWCleanup?

Aha! No!

```
/home/john/dev/build/gaffer/lib/libOpenImageIO.so.1.7(_ZN11OpenImageIO4v1_77premultEiiiiiiNS0_8TypeDescEPvlllii+0x20b)[0x7f92f48b92eb]
/home/john/dev/build/gaffer/lib/libOpenImageIO.so.1.7(_ZN11OpenImageIO4v1_79TIFFInput14read_scanlinesEiiiiiNS0_8TypeDescEPvll+0x135)[0x7f92f4808e85]
./lib/libIECoreImage.so(_ZN11IECoreImage11ImageReader14Implementation16readTypedChannelIfEEN5boost13intrusive_ptrIN6IECore4DataEEEmN11OpenImageIO4v1_78TypeDescE+0x223)[0x7f92f52013c3]
./lib/libIECoreImage.so(_ZN11IECoreImage11ImageReader14Implementation11readChannelERKSsb+0xb4)[0x7f92f5204b14]
./lib/libIECoreImage.so(_ZN11IECoreImage11ImageReader11doOperationEPKN6IECore14CompoundObjectE+0x1e2)[0x7f92f51ffe62]
```

So there the crash is in `OpenImageIO::v1_7::premult`. What's more...

```
> tiffinfo crash.tif
...
Extra Samples: 1<unassoc-alpha>
...
> tiffinfo noCrash.tif
...
Extra Samples: 1<assoc-alpha>
...
```
The two tiffs in this case are the before and after versions of test/IECoreGL/images/bSplineCircle.tif. The main difference seems to be that the bad one is unpremultiplied, and the good one is premultiplied. So presumably only the bad one hits that `OpenImageIO::v1_7::premult` code path. Might be something to investigate?

> Also, we just might have OIIO 1.8 installed here soon, and potentially OSL after that, so if you want to send a link to your Gaffer-uses-this-stuff branch, I can try that out here as well. If its not ready for PR yet, just send it privately.

Sure thing.

andrewkaufman commented:
I've pushed 4 new commits now. fe25789 reverts the tiff re-write, and a015de8 reimplements the ImageReader such that those tiffs no longer cause crashes, and their tests pass.

However, at the same time I upgraded to OIIO 1.8.4dev at IE, and that required 2 more commits. 651f497 is harmless since the `OIIO` namespace exists in OIIO 1.7 and 1.8. But b8876c3 is a bit worrying... visual verification of those test images shows they are a bit different now, though the differences are quite subtle. This seems to have been the case prior to a015de8 though, so I'm wondering if its an OIIO change this time round...

johnhaddon commented on src/IECoreImage/ImageReader.cpp:
```diff
@@ -343,34 +326,44 @@ class ImageReader::Implementation
-			return true;
+		static void destroyImageCache( ImageCache *cache )
+		{
+			ImageCache::destroy( cache, /* teardown */ true );
```
The docs (and source) for OIIO seem to suggest that this teardown argument isn't relevant when the cache isn't shared anyway, so can we just use `ImageCache::destroy` directly rather than wrap it?

andrewkaufman replied:
When I tried to do it directly:

`std::unique_ptr<ImageCache, decltype(&ImageCache::destroy) > m_cache;`

I got this compile error:

```
src/IECoreImage/ImageReader.cpp:369:60: error: decltype cannot resolve address of overloaded function
   std::unique_ptr<ImageCache, decltype(&ImageCache::destroy) > m_cache;
```

The wrap seemed simple enough, unless you know the necessary c++11 shenanigans to get this to compile without?

Also, wrapping protects us in case the relevance of the teardown argument changes in the future.

johnhaddon replied:
> The wrap seemed simple enough, unless you know the necessary c++11 shenanigans to get this to compile without?

I think `std::unique_ptr<ImageCache, void (*)( ImageCache * )>` should do it.

> Also, wrapping protects us in case the relevance of the teardown argument changes in the future.

I doubt that will ever happen. In the meantime, as someone trying to understand the code, I had to jump through another level of indirection, spot that the indirection was to pass a parameter I hadn't heard of, look up the documentation for that parameter, realise that it wasn't doing anything anyway and then try to figure out why this was necessary. It's definitely clearer just to use `ImageCache::destroy` directly...

johnhaddon commented on src/IECoreImage/ImageReader.cpp:
```diff
@@ -343,34 +326,44 @@ class ImageReader::Implementation
-			if( !m_inputFile )
+			m_cache.reset( ImageCache::create( /* shared */ false ) );
+
+			if( ImageInput *input = ImageInput::create( m_reader->fileName() ) )
```
Is this necessary? It seems wrong that we're using both the ImageCache _and_ direct access to ImageInput. In Gaffer we appear to [get the file format from the image cache](https://github.com/GafferHQ/gaffer/blob/master/src/GafferImage/OpenImageIOReader.cpp#L615) itself.

andrewkaufman replied:
Oh, nice. I'll try that out.

andrewkaufman replied:
Addressed in 73be01e

johnhaddon commented on src/IECoreImage/ImageReader.cpp:
```diff
@@ -343,34 +326,44 @@ class ImageReader::Implementation
-				{
-					throw IOException( string( "Failed to open file \"" ) + m_reader->fileName() + "\". " + geterror() );
-				}
+				m_cache->add_file( m_inputFileName );
```
The docs say "There is no reason to use this method unless you are supplying a custom creator, or configuration, or both." Is there a reason other than that?

andrewkaufman replied:
Addressed in 73be01e

johnhaddon commented:
Just ran everything at my end (OIIO 1.7.15) and all is well - no crashes, all tests pass.

andrewkaufman commented:
>> But b8876c3 is a bit worrying... visual verification of those test images shows they are a bit different now, though the differences are quite subtle.

> What sort of differences are we talking about? Across-the-board-color-changes? Differences at the alpha fringes?

Actually, looking at bit closer, I'm wondering if its a colorspace thing maybe. I was tricked into thinking meshST.tif was nearly identical because RV applies color correction to the image individually...

andrewkaufman commented:
Although, do the test still pass for you with OIIO 1.7.15 even if you revert b8876c3 ?

johnhaddon commented:
> Although, do the test still pass for you with OIIO 1.7.15 even if you revert b8876c3 ?

Ah, no, with that reverted I get these failures :

```
======================================================================
FAIL: test (DiskPrimitiveTest.DiskPrimitiveTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/disk1/john/dev/cortex/test/IECoreGL/DiskPrimitiveTest.py", line 88, in test
    self.assertEqual( a[i], 0 )
AssertionError: 1.0 != 0

======================================================================
FAIL: testVertexAttributes (MeshPrimitiveTest.MeshPrimitiveTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/disk1/john/dev/cortex/test/IECoreGL/MeshPrimitiveTest.py", line 105, in testVertexAttributes
    self.assertEqual( IECoreImage.ImageDiffOp()( imageA = expectedImage, imageB = actualImage, maxError = 0.05 ).value, False )
AssertionError: True != False

======================================================================
FAIL: testVertexAttributes (PointsPrimitive.TestPointsPrimitive)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/disk1/john/dev/cortex/test/IECoreGL/PointsPrimitive.py", line 135, in testVertexAttributes
    self.assertEqual( IECoreImage.ImageDiffOp()( imageA = expectedImage, imageB = actualImage, maxError = 0.05 ).value, False )
AssertionError: True != False

======================================================================
FAIL: testSprites (TextTest.TextTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/disk1/john/dev/cortex/test/IECoreGL/TextTest.py", line 120, in testSprites
    self.assertEqual( IECoreImage.ImageDiffOp()( imageA=imageCreated, imageB=expectedImage, maxError=0.0575 ), IECore.BoolData( False ) )
AssertionError: IECore.BoolData( 1 ) != IECore.BoolData( 0 )

----------------------------------------------------------------------
```

I guess that rules out OIIO changes between 1.7 and 1.8 being to blame...

andrewkaufman commented:
Looks like in 6b494b9b916fe06ed4a523ed4549e8da97853c71 I had removed a setting indicating the expectedResult images in `DiskPrimitiveTest`, `MeshPrimitiveTest`, and `TestPointsPrimitive` were already linear despite being tiffs. Reading rawChannels here lets those 3 tests pass with the existing threshold.

For `testSprites`, RV indicated that the expectedResults image was also linear, so I've added the same rawChannels option to that one, despite it not being there prior to this PR, and that now passes as well. That test had been failing for a good while, so I expect it should have been there all along.

I did a rebase to drop the tiff re-write & revert commits, as well as the GL tollerance tweak & revert. The only new commits since @johnhaddon's last comments are 30d3c151b75882c5d353d1be076fd7421da153c3, 43701458202b1cce4812cc3def27647e31b91f69, and 8f6e5d6d0ec198baa9ff58d8ed84b7f8351bf620.

johnhaddon commented:
I'm still fixing up a bunch of places in the Gaffer tests where we were relying on ImagePrimitiveEvaluator, but other than that I think we're good to go. I did just notice though that DisplayDriver and DisplayDriverServer etc are still in IECore - is there any reason for that? Perhaps you could move them over too?

andrewkaufman commented:
> I did just notice though that DisplayDriver and DisplayDriverServer etc are still in IECore - is there any reason for that?

Just that we didn't list them in the headers-to-move spreadsheet. I can try moving them now, hopefully it will go smoothly.

andrewkaufman commented:
Alright, rebased on master for hopefully the last time. The last 7 commits are new, the rest remain unchanged since the previous review. 5 of those commits are addressing the request to move `DisplayDriver` and associated classes to IECoreImage.

88c1dfc is a minor IE build issue that is now fixed.

1c14326 is a fairly major IE build issue that I don't have a good solution for. The build feature seemed reasonable anyways though, as someone might be interested in IECoreHoudini but not want our GL drawing code.

andrewkaufman commented:
Ugg, comment updated inplace

johnhaddon commented on SConstruct:
```diff
@@ -1433,14 +1433,16 @@ if doConfigure :
-		corePythonSources.remove( "src/IECorePython/DisplayDriverServerBinding.cpp" )
-		corePythonSources.remove( "src/IECorePython/DisplayDriverBinding.cpp" )
-		corePythonSources.remove( "src/IECorePython/ImageDisplayDriverBinding.cpp" )
+		coreSources.remove( "src/IECoreImage/ClientDisplayDriver.cpp" )
```
`coreSources`only  contains `glob.glob( "src/IECore/*.cpp" )`, so this won't be removing anything. How about we just remove the checks and require asio?

andrewkaufman replied:
It seems asio is only required by IECoreImage and IECoreNuke. Should I make it a primary requirement incase we want it for IECore in the future, or leave it to those libs only?

andrewkaufman replied:
Similarly, -`DIECORE_WITH_SIGNALS` doesn't seemed to be used for anything, so maybe I should remove and make that a hard requirement as well?

johnhaddon replied:
I think probably just add it to the LIBS for IECoreImage and IECoreNuke, but don't worry about trying to compile them with bits missing if it isn't found. Just assume it'll be there basically, since we've already done a general "is boost available?" check earlier on.

johnhaddon replied:
> Similarly, -DIECORE_WITH_SIGNALS doesn't seemed to be used for anything, so maybe I should remove and make that a hard requirement as well?

Yeah, all this dates way back to when those libraries were first introduced to boost (somewhere in the 1.3x range I think). VFXPlatform version is 1.61 so it's all ancient history...

andrewkaufman replied:
So at this points, freetype will be the only "optional" dependency of IECore. Should I just make that required too?

johnhaddon replied:
Let's leave freetype as-is, at least until that dependency is moved out from the main lib.

andrewkaufman replied:
Addressed in 94e0725

johnhaddon commented:
Thanks!
